A variety of edits for the paper

marcharper opened the pull request:
Some suggestions:

The point of figure 6 needs to be more clear. At a glance the figure seems to say that our results don't match theory, which is not the case.
 
We should use a different color for dotted neutral lines in the boxplot charts

Move figure 14 to start of invasion section and figure 16 to start of resistance section. These give the most information and are the most interesting IMO. If the paper were nothing but two diagrams summarizing the results I would want it to be these two.

Fig 14-16: grid lines make some strategies hard to see -- perhaps an all white background is better.

Can tables 3 and 4 be simplified? There is a lot of redundant information. Perhaps a separate table listing the memory depth and stochasticity of all strategies mentioned in the paper would be a good idea. However I understand the desire to point out that the strategies are not complex.

Table 5: We should insert lines between the good invaders, good resistors, and good ZD strategies.

We may have taken the third-person approach a little too far: 
"In this section the focus is on the ability of a mutant strategy to invade" might be better as "We now focus on the ability of a mutant to invade". 

The size of the tables of invasion and fixation are hiding something about TF*. True they are not in the top 5 invaders but they are 7, 9, and 14 for N=2 and TF3 is rank 16 for N=14 (top 10%). They are also all near the top for resistance for and N/2. Other than CS and Evolved LookerUp 1_1_1 no other strategy is very good at both consistently, so by training at a starting state of (N/2, N/2) we got good invaders and great resistors. I recommend that we extend the top 5 tables to give a better picture.

On that topic I added a discussion section that argues that we can't really do better unless players know the population distribution (or you had a tag with the opponent's type). Let me know if that section doesn't make sense. I convinced myself it does...

WinStayLoseShift should be abbreviated WSLS.

marcharper committed:
A variety of edits for the paper

marcharper committed:
Minor edits

drvinceknight commented on main.tex:
```diff
@@ -23,6 +23,7 @@
 % Title
 \title{An Empirical Study of Invasion and Resistance
        for Iterated Prisoner's Dilemma Strategies}
+\title{Evolving Invaders and Resistors for the Iterated Prisoner's Dilemma}
```
Nice!

drvinceknight replied:
On reflection, I'm not sure this is the correct title. We don't spend much time discussing the training of the Invaders and Resistors...

marcharper replied:
I'm open to other titles

drvinceknight replied:
@drvinceknight puts thinking cap on...

drvinceknight commented on main.tex:
```diff
@@ -78,15 +82,17 @@ \section{Introduction}\label{sec:introduction}
-strategies. Not all strategies have been considered for this study: same use
-knowledge of the game such as the number of turns and others have a high
+strategies. Not all strategies have been considered for this study: excluded
+are those that make use of knowledge of the game such as the number of turns
```
This is a bit misleading, all our ZD strategies make use of the values of the game right? The only ones we've removed are the ones that make use of the length of the game.

drvinceknight commented on main.tex:
```diff
@@ -335,23 +344,23 @@ \section{Validation}\label{sec:validation}
-\begin{equation}\label{equ:expected_payoff_two}
-    g_i = 1 - \omega + \omega G_i
-\end{equation}
+% With an intensity of selection \(\omega\) the fitness of both strategies is
```
Good call, let's just delete this.

drvinceknight commented on main.tex:
```diff
@@ -550,11 +552,7 @@ \subsection{The special case of \(N=2\)}\label{sec:two_individuals}
-        defect in the future it cooperates. Otherwise it defects for all
-        rounds. This strategy was specifically designed for
-        evolutionary processes \cite{Li2009}.
+        handshake mechanism described above.
```
`described in Section~\ref{sec:strategies}`

drvinceknight commented on main.tex:
```diff
@@ -594,15 +592,15 @@ \subsection{The special case of \(N=2\)}\label{sec:two_individuals}
 specific structure of strategies. The analysis carried out here makes no
 assumptions about the structure of the strategies by using actual simulation
-interactions.
+interactions. \cite{Stewart...}
```
Could you add this as a `% TODO` just to make sure we don't miss it, I guess you're wanting to add something here. The key is `Stewart2012`.

marcharper replied:
I had a different paper in mind actually, reference added

drvinceknight commented on main.tex:
```diff
@@ -594,15 +592,15 @@ \subsection{The special case of \(N=2\)}\label{sec:two_individuals}
 
-In the next sections close attention to
-strategies who are strong invaders/resistors is given.
+% In the next sections close attention to
```
Not a fan of a transition sentence between sections?

marcharper replied:
Well it's just "in the next section X" and then the next section opens with "In this section X" in some cases, which was redundant

drvinceknight replied:
Fine. :+1:

drvinceknight commented on main.tex:
```diff
@@ -905,6 +904,58 @@ \subsection{The effect of population size}\label{sec:population_size}
+of cooperation, these strategies are not the most evolutionarily stable
+in the long run. A strategy with a handshaking mechanism is still capable of
+invading and is more resistant to subsequent invasions. Moreover, the
+best resistor of the payoff maximally trained strategies (EvolvedLookerUp1_1_1),
```
This won't compile: `EvolvedLookerUp1\_1\_1`.

drvinceknight replied:
I've fixed this in #36

drvinceknight commented on main.tex:
```diff
@@ -905,6 +904,58 @@ \subsection{The effect of population size}\label{sec:population_size}
+$(N/2, N/2)$. Nevertheless our results suggest that there is not much room for
+improvement. Any handshake more sophisticated than C necessarily involves
+a defection. (A handshake consisting of a longer sequence of cooperations is
+effectively a grudger.) For TF3 or EvolvedLookerUp1_1_1 to become better resistors
```
`EvolvedLookerup1\_1\_1`

drvinceknight replied:
Fixed in #36

drvinceknight commented on main.tex:
```diff
@@ -905,6 +904,58 @@ \subsection{The effect of population size}\label{sec:population_size}
+a defection. (A handshake consisting of a longer sequence of cooperations is
+effectively a grudger.) For TF3 or EvolvedLookerUp1_1_1 to become better resistors
+they need a longer or more strict handshake. But if this handshake involves
+a defection then likely the invasion ability is diminished for $N > 2$: the top
```
`then it is likely that the invasion`? (I don't know, as is read funky to me, I might be wrong)

marcharper replied:
it is likely

drvinceknight commented on main.tex:
```diff
@@ -905,6 +904,58 @@ \subsection{The effect of population size}\label{sec:population_size}
+are only effective invaders for the smallest $N$, dropping down dramatically
+as the population size increases.
+
+We did, however, attempt to evolve CS using FSM and lookup table based players,
```
Just noting the `We` depending one what "it" is decided to do ;)

drvinceknight commented on main.tex:
```diff
@@ -951,11 +1002,11 @@ \section{Conclusion}\label{sec:conclusion}
-It would also be possible to consider three or more types of strategy in the
-population and finally mutation would also offer an interesting dimension to
-explore.
+% There are various areas for further work to build on this. Firstly, an analysis
```
Not a fan?

marcharper replied:
I meant to come back to that

drvinceknight replied:
Cool.

drvinceknight reviewed (changes requested):
This all looks really great @marcharper, I've made a couple of very minor suggestions for changes.

drvinceknight commented:
> We may have taken the third-person approach a little too far:
> "In this section the focus is on the ability of a mutant strategy to invade" might be better as "We now focus on the ability of a mutant to invade".

I'd stick with the first if this was only me but that's purely based on a really painful review a while back (ridiculous back and forth that escalated over this petty stuff) after which I just decided to stick with third person. I also find when working collaboratively it helps because it avoids confusion between we, I etc... I'm easy if we want to change though (if we happen to get a silly reviewer it's an easy enough thing to change back). I pointed out some "We"s in your changes just in case we do go with third person.

drvinceknight commented:
> On that topic I added a discussion section that argues that we can't really do better unless players know the population distribution (or you had a tag with the opponent's type). Let me know if that section doesn't make sense. I convinced myself it does...

It made sense to me. :+1: I thought the discussion section was great actually.

drvinceknight committed:
Improve plots
Addresses #34.

- Change color for neutral fixation in boxplots. Gone from blue to black.
- Remove grid lines in the rank plots. Still have the grey background
  but we can change that.
- Move the invasion and resistance rank plots to the corresponding
  sections.

drvinceknight committed:
Improve tables
Addresses #35

- Rewrite appendix with player information.
- Simplify the tables of top performing strategies.
- Present top 10% of strategies as opposed to top 5.
- Make TF* appear in bold.
- Include lines to separate strategies in rank tables and also display ints

drvinceknight committed:
Make minor adjustments.

drvinceknight committed:
Rerun all graphics.

marcharper committed:
Merge pull request #36 from Axelrod-Python/34

Adjustments based on comments in #33

drvinceknight commented:
> The point of figure 6 needs to be more clear. At a glance the figure seems to say that our results don't match theory, which is not the case.

Can you expand on this a bit? My understanding was that figure 6 shows the shortcoming of the theory which is based on a particular theoretic result based on the average payoff.

marcharper commented:
It's a shortcoming of using average payoff to compute the transition probabilities. We could compute the transition probabilities directly, in which case the formula that uses only the transition probabilities would be correct. Given the transition probabilities the fixation probability (in the two type case) can be computed exactly.

The issue is that it is difficult to come up with a formula for an arbitrary strategy because the fitness landscape is difficult to compute for strategies with memory depth > 1, for arbitrary match lengths, etc. The formula doesn't apply to repeated games (rather it's for one-shot games), that's why previous authors used the stationary score -- it reduces a match to a single score if the stationary distribution can be computed (and there are infinitely many rounds).

drvinceknight commented:
> The issue is that it is difficult to come up with a formula for an arbitrary strategy because the fitness landscape is difficult to compute for strategies with memory depth > 1, for arbitrary match lengths, etc. The formula doesn't apply to repeated games (rather it's for one-shot games), that's why previous authors used the stationary score -- it reduces a match to a single score if the stationary distribution can be computed (and there are infinitely many rounds).

That makes sense, I'll open a PR with a suggested edit around that but feel free to just ignore it and write what you feel is best. :+1:

drvinceknight committed:
Add explanation about validation for stochastic.

Discussed on #33

drvinceknight commented:
https://github.com/Axelrod-Python/axelrod-moran/pull/37

drvinceknight committed:
Resize the boxplots.

This puts them all next to each other and also has the benefit of
increasing the font size.

drvinceknight commented:
Have opened #38 which resizes the boxplots to fit on a single page, this looks good for the multiple ones for invasion and resistance but a tiny bit funky for the lonely plot for N=2 (but I still think it's an improvement over status quo).

drvinceknight committed:
Add a small discussion for longer ranking tables

drvinceknight commented:
https://github.com/Axelrod-Python/axelrod-moran/pull/39 adds a bit more discussion around the longer tables, interestingly for the reistors even within the top 16 strategies: none of them are stochastic.

marcharper committed:
Merge pull request #38 from Axelrod-Python/resize-figs

Resize the boxplots.

marcharper committed:
Requested edits

marcharper committed:
Correct Moran process algorithm chart

marcharper committed:
More updates

marcharper committed:
Update abstract

marcharper commented on main.tex:
```diff
@@ -108,13 +131,17 @@ \section{Introduction}\label{sec:introduction}
+ and achieve mutual cooperation with many other strategies.
 \end{enumerate}
 
+These finding have the potential to offer insight in to organisms such as antibacterial
```
I'm not sure if I understand the connection to antibacterial resistance.

drvinceknight replied:
My understanding of antibacterial resistant bacteria, is that they are now a huge problem because of the fact that they have evolved to resist antibiotics: they are in effect organisms that have evolved to be strong at resisting invasion.

I saw a strong parallel with what we're observing here with our strategies: and what I'm postulating is that these resistant bacteria would have some sort of handshaking mechanism (similar to our strategies). 

It's a bit of a step but it's what the evolution of our handshakes made me think about (the resistant bacteria have evolved "something" to become a big problem).

marcharper replied:
It's a different kind of resistance I think -- resistance to an environmental component rather than invading variants like with a handshake. Handshaking is more like chemical signaling between organisms, tribal fashion, etc.

drvinceknight replied:
Damn that's a pity :) Let's merge this PR and I'll add something to remove it.

marcharper committed:
Conclusion paragraph

marcharper committed:
Authors TODO

drvinceknight committed:
Merge pull request #37 from Axelrod-Python/explain-fig-6

Add explanation about validation for stochastic.

drvinceknight committed:
Make amendments suggested by Marc.

drvinceknight committed:
Rephrase sentence about the removed strategies.

marcharper committed:
Merge pull request #40 from Axelrod-Python/removed-strategies

Rephrase sentence about the removed strategies.

marcharper committed:
Merge pull request #39 from Axelrod-Python/discuss-more-strategies

Add a small discussion for longer ranking tables
